Modules: Deduplicate suffixes but don't sort them.

RemiLacroix-IDRIS opened the pull request:
PR #14920 changed the behavior to sorting the suffixes. That change is causing some issues with our modules.

`foo/X.Y.Z-cuda-mpi` becomes `foo/X.Y.Z-cuda-mpi` for example, which we aren't really happy about because it is causing duplicated modules if we don't pay attention and we cannot have consistent module naming without breaking existing submission scripts.

The documentation (https://spack.readthedocs.io/en/latest/module_file_support.html#selection-by-anonymous-specs) currently states:
> Order does matter
> The modifications associated with the all keyword are always evaluated first, no matter where they appear in the configuration file. All the other spec constraints are instead evaluated top to bottom.

I think that it makes sense that "modifications" also apply to `suffixes`. As far as I can tell, the suffixes' ordering is indeed guaranteed to follow the ordering of configuration files (except for `all` which is considered first, as already noted in the documentation).

RemiLacroix-IDRIS committed:
Modules: Deduplicate suffixes but don't sort them.

The suffixes' order is defined by the order in which they appear in the configuration file.

alalazo reviewed (changes requested):
This change in behavior was part of the [0.14.1 Changelog](https://github.com/spack/spack/releases/tag/v0.14.1) and this PR reverts it. I would advocate for not bouncing back and forth with behavior, especially considering that [this](https://spackpm.slack.com/archives/C5WF9RXPZ/p1582117727074200) was posted at the time. Can you be more specific on the issues that this causes for you? Wondering if it is possible to find an alternative solution.

RemiLacroix-IDRIS commented:
I am sorry I missed that change earlier but it is hard to keep track, especially when you let some time go by without updating your working Spack install...

> Can you be more specific on the issues that this causes for you? Wondering if it is possible to find an alternative solution.

If we refresh an existing module `foo/X.Y.Z-mpi-cuda`, it will create a duplicated module `foo/X.Y.Z-cuda-mpi`.

If we create a module for a new product, it will create a module `foo/X.Y.Z-cuda-mpi` while we may already have `foo/I.J.K-mpi-cuda`, leading to inconsistent module names.

I cannot think of any solution that would allow forcing the order to `-mpi-cuda`. Renaming existing module files is not something we would be happy to do as users HATE when that sort of things change as it breaks their submission scripts.

alalazo commented:
> Renaming existing module files is not something we would be happy to do as users HATE when that sort of things change as it breaks their submission scripts.

Are you updating in place the Spack that is used for site-wide deployment? Would it be a solution to generate a new tree of modules at another location and provide users with a command to swap between the two?

RemiLacroix-IDRIS commented:
> Are you updating in place the Spack that is used for site-wide deployment?

Yeah, we already did in fact. Unfortunately I missed that issue when updating and I am only noticing it now, a few weeks after the upgrade. :/

> Would it be a solution to generate a new tree of modules at another location and provide users with a command to swap between the two?

I don't really think we want to go down that road. To be honest, I think that if this PR is rejected, we will just keep the patch for our install, it will be easier for us.

Additionally I do think it is nice to be able to control the ordering of the suffixes.

sethrj commented:
@RemiLacroix-IDRIS I think you might also have missed [this comment](https://github.com/spack/spack/pull/14920#issuecomment-585810247): at least at the time of that pull request, the ordering of the suffixes was unspecified: python 3.6+ kept them in the order of creation, but other versions of python shuffled them around. So the changeset you've submitted might not even get you the behavior you want.

Indeed, given that the new test I added (https://github.com/spack/spack/pull/14920/files) doesn't change behavior in your PR, I suspect this isn't doing what you want.

sethrj commented:
P.S.  Is the double-hyphen in your PR description `foo/X.Y.Z--cuda-mpi` a typo or actually a generated module name? Because that *does*  look like a bug that should be fixed.

RemiLacroix-IDRIS commented:
> P.S. Is the double-hyphen in your PR description `foo/X.Y.Z--cuda-mpi` a typo or actually a generated module name? Because that _does_ look like a bug that should be fixed.

That's a typo (now fixed).

RemiLacroix-IDRIS commented:
> @RemiLacroix-IDRIS I think you might also have missed [this comment](https://github.com/spack/spack/pull/14920#issuecomment-585810247): at least at the time of that pull request, the ordering of the suffixes was unspecified: python 3.6+ kept them in the order of creation, but other versions of python shuffled them around. So the changeset you've submitted might not even get you the behavior you want.

I had seen it but I cannot understand how it would be possible that the ordering is different, unless the ordering of `ordereddict` changed somewhere along the way.

> Indeed, given that the new test I added (https://github.com/spack/spack/pull/14920/files) doesn't change behavior in your PR, I suspect this isn't doing what you want.

That might not have been what was intended back then but `bar-foo` is what is expected when using the config file order and when sorting.

sethrj reviewed (changes requested):
This PR needs a test demonstrating the change in behavior. See lib/spack/spack/test/modules/tcl.py . Because the old test for sorted behavior is passing, I believe your change isn't doing what you think it's doing.

RemiLacroix-IDRIS commented:
> @RemiLacroix-IDRIS I think you might also have missed [this comment](https://github.com/spack/spack/pull/14920#issuecomment-585810247): at least at the time of that pull request, the ordering of the suffixes was unspecified: python 3.6+ kept them in the order of creation, but other versions of python shuffled them around. So the changeset you've submitted might not even get you the behavior you want.

You are correct, it seems that there are some remaining `dict` in the config, although Spack seems to use `ordereddict` mostly everywhere.

Turns out that in Python 3.6+ `dict` implementation is ordered so everything works as expected.

I am looking into this to understand whether it is expected or not that `dict` are used.

> Indeed, given that the new test I added (https://github.com/spack/spack/pull/14920/files) doesn't change behavior in your PR, I suspect this isn't doing what you want.

That might not have been what was intended back then but as far as I can tell `bar-foo` is what is expected both when using the config file order and when sorting.

If I modify the test so that the order is different, I can see that I don't get the expected behavior with Python 2.7 after my change.

sethrj commented:
@RemiLacroix-IDRIS Before you go any further, I'd recommend double-checking that retaining the original order actually updates the module names to what you expect from your original module name generation. It's possible that if the extensions are arbitrarily ordered there's nothing that can be done.

RemiLacroix-IDRIS committed:
Modules: Modify tests to use spack_yaml.load_config.

spack_yaml.load_config ensures that the configuration is stored in an ordered manner. Without this change, the behavior of the tests did not match Spack's.

RemiLacroix-IDRIS committed:
Modules: Tweak the suffixes test to better catch ordering issues.

RemiLacroix-IDRIS commented:
> Before you go any further, I'd recommend double-checking that retaining the original order actually updates the module names to what you expect from your original module name generation. It's possible that if the extensions are arbitrarily ordered there's nothing that can be done.

It does match the original module name generation when using the config file order.

> I am looking into this to understand whether it is expected or not that dict are used.

Turns out the problem was in the tests themselves! The configuration file was loaded differently in Spack and in the modules tests. Spack does use ordered dict everywhere, even with Python 2.7!

sethrj reviewed (approved):
This looks good to me with the added tests (good catch with the unit tests being different from the main code!). @alalazo we may want to post another Spack survey/announcement making sure no one else will be taken by surprise by this change.

alalazo commented:
@sethrj I agree with posting a comment in #modules and merge this PR if no objection arise in 1-2 weeks. Can you post that message? 

@RemiLacroix-IDRIS  Do you have time to add a few more tests to check that the order of suffixes is the one we expect?

RemiLacroix-IDRIS commented:
> Do you have time to add a few more tests to check that the order of suffixes is the one we expect?

@alalazo What do you have in mind exactly? It's what I tried to do with 336283806f0b9e567591ecb3ec1ce0ce9afc979e.

alalazo commented:
@RemiLacroix-IDRIS For instance, if I understand this changeset correctly, maintaining the original configuration for the test we should have `bar-foo` as a suffix, right? I think we should check both and maybe other edge cases to verify the semantics are correct.

RemiLacroix-IDRIS commented:
@alalazo I have extended the tests a bit.

sethrj commented:
OK, it's been a week with one positive and no negative feedback. The old (sorted) behavior is easy to do (just sort the keywords in your module file), so it should be easy for anyone to work around if they're surprised by the updated behavior. Merging!
